// @douyinfe/semi-illustrations/lib/cjs/illustrations/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var illustrations_exports = {};
__export(illustrations_exports, {
  IllustrationConstruction: () => import_IllustrationConstruction.default,
  IllustrationConstructionDark: () => import_IllustrationConstructionDark.default,
  IllustrationFailure: () => import_IllustrationFailure.default,
  IllustrationFailureDark: () => import_IllustrationFailureDark.default,
  IllustrationIdle: () => import_IllustrationIdle.default,
  IllustrationIdleDark: () => import_IllustrationIdleDark.default,
  IllustrationNoAccess: () => import_IllustrationNoAccess.default,
  IllustrationNoAccessDark: () => import_IllustrationNoAccessDark.default,
  IllustrationNoContent: () => import_IllustrationNoContent.default,
  IllustrationNoContentDark: () => import_IllustrationNoContentDark.default,
  IllustrationNoResult: () => import_IllustrationNoResult.default,
  IllustrationNoResultDark: () => import_IllustrationNoResultDark.default,
  IllustrationNotFound: () => import_IllustrationNotFound.default,
  IllustrationNotFoundDark: () => import_IllustrationNotFoundDark.default,
  IllustrationSuccess: () => import_IllustrationSuccess.default,
  IllustrationSuccessDark: () => import_IllustrationSuccessDark.default
});
module.exports = __toCommonJS(illustrations_exports);
var import_IllustrationConstructionDark = __toESM(require("./IllustrationConstructionDark"));
var import_IllustrationConstruction = __toESM(require("./IllustrationConstruction"));
var import_IllustrationFailureDark = __toESM(require("./IllustrationFailureDark"));
var import_IllustrationFailure = __toESM(require("./IllustrationFailure"));
var import_IllustrationIdleDark = __toESM(require("./IllustrationIdleDark"));
var import_IllustrationIdle = __toESM(require("./IllustrationIdle"));
var import_IllustrationNoAccessDark = __toESM(require("./IllustrationNoAccessDark"));
var import_IllustrationNoAccess = __toESM(require("./IllustrationNoAccess"));
var import_IllustrationNoContentDark = __toESM(require("./IllustrationNoContentDark"));
var import_IllustrationNoContent = __toESM(require("./IllustrationNoContent"));
var import_IllustrationNoResultDark = __toESM(require("./IllustrationNoResultDark"));
var import_IllustrationNoResult = __toESM(require("./IllustrationNoResult"));
var import_IllustrationNotFoundDark = __toESM(require("./IllustrationNotFoundDark"));
var import_IllustrationNotFound = __toESM(require("./IllustrationNotFound"));
var import_IllustrationSuccessDark = __toESM(require("./IllustrationSuccessDark"));
var import_IllustrationSuccess = __toESM(require("./IllustrationSuccess"));
