// @douyinfe/semi-foundation/lib/es/resizable/singleConstants.js
// single
const rowStyleBase = {
  width: '100%',
  height: '10px',
  top: '0px',
  left: '0px',
  cursor: 'row-resize'
};
const colStyleBase = {
  width: '10px',
  height: '100%',
  top: '0px',
  left: '0px',
  cursor: 'col-resize'
};
const edgeStyleBase = {
  width: '20px',
  height: '20px',
  position: 'absolute'
};
export const directions = ['top', 'right', 'bottom', 'left', 'topRight', 'bottomRight', 'bottomLeft', 'topLeft'];
export const directionStyles = {
  top: Object.assign(Object.assign({}, rowStyleBase), {
    top: '-5px'
  }),
  right: Object.assign(Object.assign({}, colStyleBase), {
    left: undefined,
    right: '-5px'
  }),
  bottom: Object.assign(Object.assign({}, rowStyleBase), {
    top: undefined,
    bottom: '-5px'
  }),
  left: Object.assign(Object.assign({}, colStyleBase), {
    left: '-5px'
  }),
  topRight: Object.assign(Object.assign({}, edgeStyleBase), {
    right: '-10px',
    top: '-10px',
    cursor: 'ne-resize'
  }),
  bottomRight: Object.assign(Object.assign({}, edgeStyleBase), {
    right: '-10px',
    bottom: '-10px',
    cursor: 'se-resize'
  }),
  bottomLeft: Object.assign(Object.assign({}, edgeStyleBase), {
    left: '-10px',
    bottom: '-10px',
    cursor: 'sw-resize'
  }),
  topLeft: Object.assign(Object.assign({}, edgeStyleBase), {
    left: '-10px',
    top: '-10px',
    cursor: 'nw-resize'
  })
};
export const DEFAULT_SIZE = {
  width: 'auto',
  height: 'auto'
};
