// @douyinfe/semi-ui/lib/es/resizable/group/resizeHandler.js
import React, { createRef } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { ResizeHandlerFoundation } from '@douyinfe/semi-foundation/lib/es/resizable/foundation';
import { cssClasses } from '@douyinfe/semi-foundation/lib/es/resizable/constants';
import { directionStyles } from '@douyinfe/semi-foundation/lib/es/resizable/groupConstants';
import BaseComponent from '../../_base/baseComponent';
import { ResizeContext } from './resizeContext';
import { IconHandle } from '@douyinfe/semi-icons';
const prefixCls = cssClasses.PREFIX;
class ResizeHandler extends BaseComponent {
  constructor(props) {
    super(props);
    this.onMouseDown = e => {
      const {
        notifyResizeStart
      } = this.context;
      notifyResizeStart(this.handlerIndex, e);
    };
    this.getHandler = () => {
      return this.handlerRef.current;
    };
    this.state = {};
    this.handlerRef = /*#__PURE__*/createRef();
    this.foundation = new ResizeHandlerFoundation(this.adapter);
  }
  componentDidMount() {
    this.foundation.init();
    this.handlerIndex = this.context.registerHandler(this.handlerRef);
  }
  componentDidUpdate(_prevProps) {}
  componentWillUnmount() {
    this.foundation.destroy();
  }
  get adapter() {
    return Object.assign(Object.assign({}, super.adapter), {
      registerEvents: () => {
        this.handlerRef.current.addEventListener('mousedown', this.onMouseDown);
      },
      unregisterEvents: () => {
        this.handlerRef.current.removeEventListener('mousedown', this.onMouseDown);
      }
    });
  }
  render() {
    const {
      style,
      className,
      children
    } = this.props;
    return /*#__PURE__*/React.createElement("div", {
      className: classNames(className, prefixCls + '-handler'),
      style: Object.assign(Object.assign({}, directionStyles[this.context.direction]), style),
      ref: this.handlerRef
    }, children !== null && children !== void 0 ? children : /*#__PURE__*/React.createElement(IconHandle, {
      size: 'inherit',
      style: {
        rotate: this.context.direction === 'horizontal' ? '0deg' : '90deg'
      }
    }));
  }
}
ResizeHandler.propTypes = {
  children: PropTypes.node,
  direction: PropTypes.string,
  onResizeStart: PropTypes.func,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  style: PropTypes.object
};
ResizeHandler.defaultProps = {};
ResizeHandler.contextType = ResizeContext;
export default ResizeHandler;
