// @douyinfe/semi-ui/lib/es/resizable/group/resizeItem.js
import _noop from "lodash/noop";
import React, { createRef } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { ResizeItemFoundation } from '@douyinfe/semi-foundation/lib/es/resizable/foundation';
import { cssClasses } from '@douyinfe/semi-foundation/lib/es/resizable/constants';
import BaseComponent from '../../_base/baseComponent';
import { ResizeContext } from './resizeContext';
const prefixCls = cssClasses.PREFIX;
class ResizeItem extends BaseComponent {
  constructor(props) {
    super(props);
    this.itemRef = /*#__PURE__*/createRef();
    this.foundation = new ResizeItemFoundation(this.adapter);
    this.state = {
      isResizing: false
    };
  }
  componentDidMount() {
    this.foundation.init();
    const {
      min,
      max,
      onResizeStart,
      onChange,
      onResizeEnd,
      defaultSize
    } = this.props;
    this.itemIndex = this.context.registerItem(this.itemRef, min, max, defaultSize, onResizeStart, onChange, onResizeEnd);
  }
  componentDidUpdate(_prevProps) {}
  componentWillUnmount() {
    this.foundation.destroy();
  }
  get adapter() {
    return Object.assign({}, super.adapter);
  }
  render() {
    const style = Object.assign({}, this.props.style);
    return /*#__PURE__*/React.createElement("div", {
      style: style,
      className: classNames(this.props.className, prefixCls + '-item'),
      ref: this.itemRef
    }, this.props.children);
  }
}
ResizeItem.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
  min: PropTypes.string,
  max: PropTypes.string,
  children: PropTypes.object,
  onResizeStart: PropTypes.func,
  onChange: PropTypes.func,
  onResizeEnd: PropTypes.func,
  defaultSize: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};
ResizeItem.defaultProps = {
  onResizeStart: _noop,
  onChange: _noop,
  onResizeEnd: _noop
};
ResizeItem.contextType = ResizeContext;
export default ResizeItem;
