// @douyinfe/semi-illustrations/lib/es/illustrations/index.js
import { default as default2 } from "./IllustrationConstructionDark";
import { default as default3 } from "./IllustrationConstruction";
import { default as default4 } from "./IllustrationFailureDark";
import { default as default5 } from "./IllustrationFailure";
import { default as default6 } from "./IllustrationIdleDark";
import { default as default7 } from "./IllustrationIdle";
import { default as default8 } from "./IllustrationNoAccessDark";
import { default as default9 } from "./IllustrationNoAccess";
import { default as default10 } from "./IllustrationNoContentDark";
import { default as default11 } from "./IllustrationNoContent";
import { default as default12 } from "./IllustrationNoResultDark";
import { default as default13 } from "./IllustrationNoResult";
import { default as default14 } from "./IllustrationNotFoundDark";
import { default as default15 } from "./IllustrationNotFound";
import { default as default16 } from "./IllustrationSuccessDark";
import { default as default17 } from "./IllustrationSuccess";
export {
  default3 as IllustrationConstruction,
  default2 as IllustrationConstructionDark,
  default5 as IllustrationFailure,
  default4 as IllustrationFailureDark,
  default7 as IllustrationIdle,
  default6 as IllustrationIdleDark,
  default9 as IllustrationNoAccess,
  default8 as IllustrationNoAccessDark,
  default11 as IllustrationNoContent,
  default10 as IllustrationNoContentDark,
  default13 as IllustrationNoResult,
  default12 as IllustrationNoResultDark,
  default15 as IllustrationNotFound,
  default14 as IllustrationNotFoundDark,
  default17 as IllustrationSuccess,
  default16 as IllustrationSuccessDark
};
