// @douyinfe/semi-ui/lib/cjs/form/field.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormUpload = exports.FormTreeSelect = exports.FormTimePicker = exports.FormTextArea = exports.FormTagInput = exports.FormSwitch = exports.FormSlider = exports.FormSelect = exports.FormRating = exports.FormRadioGroup = exports.FormRadio = exports.FormInputNumber = exports.FormInput = exports.FormDatePicker = exports.FormCheckboxGroup = exports.FormCheckbox = exports.FormCascader = exports.FormAutoComplete = void 0;
var _withField = _interopRequireDefault(require("./hoc/withField"));
var _index = _interopRequireDefault(require("../input/index"));
var _textarea = _interopRequireDefault(require("../input/textarea"));
var _index2 = _interopRequireDefault(require("../inputNumber/index"));
var _index3 = _interopRequireDefault(require("../select/index"));
var _index4 = require("../checkbox/index");
var _checkboxGroup = _interopRequireDefault(require("../checkbox/checkboxGroup"));
var _index5 = require("../radio/index");
var _radioGroup = _interopRequireDefault(require("../radio/radioGroup"));
var _index6 = _interopRequireDefault(require("../datePicker/index"));
var _index7 = _interopRequireDefault(require("../switch/index"));
var _index8 = _interopRequireDefault(require("../slider/index"));
var _index9 = _interopRequireDefault(require("../timePicker/index"));
var _index10 = _interopRequireDefault(require("../treeSelect/index"));
var _index11 = _interopRequireDefault(require("../cascader/index"));
var _index12 = _interopRequireDefault(require("../rating/index"));
var _index13 = _interopRequireDefault(require("../autoComplete/index"));
var _index14 = _interopRequireDefault(require("../upload/index"));
var _index15 = _interopRequireDefault(require("../tagInput/index"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// Basic component

const FormInput = exports.FormInput = (0, _withField.default)(_index.default, {
  maintainCursor: true
});
const FormInputNumber = exports.FormInputNumber = (0, _withField.default)(_index2.default, {
  maintainCursor: true
});
const FormTextArea = exports.FormTextArea = (0, _withField.default)(_textarea.default, {
  maintainCursor: true
});
const FormSelect = exports.FormSelect = (0, _withField.default)(_index3.default);
// Select after withField is a new Component, without the Option attribute, it needs to be manually assigned once
FormSelect.Option = _index3.default.Option;
FormSelect.OptGroup = _index3.default.OptGroup;
const FormCheckboxGroup = exports.FormCheckboxGroup = (0, _withField.default)(_checkboxGroup.default);
const FormCheckbox = exports.FormCheckbox = (0, _withField.default)(_index4.Checkbox, {
  valueKey: 'checked',
  valuePath: 'target.checked',
  shouldInject: false
});
const FormRadioGroup = exports.FormRadioGroup = (0, _withField.default)(_radioGroup.default, {
  valuePath: 'target.value'
});
const FormRadio = exports.FormRadio = (0, _withField.default)(_index5.Radio, {
  valueKey: 'checked',
  valuePath: 'target.checked',
  shouldInject: false
});
const FormDatePicker = exports.FormDatePicker = (0, _withField.default)(_index6.default);
const FormSwitch = exports.FormSwitch = (0, _withField.default)(_index7.default, {
  valueKey: 'checked'
});
const FormSlider = exports.FormSlider = (0, _withField.default)(_index8.default);
const FormTimePicker = exports.FormTimePicker = (0, _withField.default)(_index9.default);
const FormTreeSelect = exports.FormTreeSelect = (0, _withField.default)(_index10.default);
const FormCascader = exports.FormCascader = (0, _withField.default)(_index11.default);
const FormRating = exports.FormRating = (0, _withField.default)(_index12.default);
const FormAutoComplete = exports.FormAutoComplete = (0, _withField.default)(_index13.default, {
  valueKey: 'value',
  onKeyChangeFnName: 'onChange'
});
const FormUpload = exports.FormUpload = (0, _withField.default)(_index14.default, {
  valueKey: 'fileList',
  valuePath: 'fileList',
  onKeyChangeFnName: 'onChange'
});
const FormTagInput = exports.FormTagInput = (0, _withField.default)(_index15.default);
