// @douyinfe/semi-ui/lib/cjs/form/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Form: true,
  ArrayField: true,
  useFormApi: true,
  useFormState: true,
  useFieldState: true,
  useFieldApi: true,
  withField: true,
  withFormState: true,
  withFormApi: true
};
Object.defineProperty(exports, "ArrayField", {
  enumerable: true,
  get: function () {
    return _arrayField.default;
  }
});
Object.defineProperty(exports, "Form", {
  enumerable: true,
  get: function () {
    return _baseForm.default;
  }
});
Object.defineProperty(exports, "useFieldApi", {
  enumerable: true,
  get: function () {
    return _index.useFieldApi;
  }
});
Object.defineProperty(exports, "useFieldState", {
  enumerable: true,
  get: function () {
    return _index.useFieldState;
  }
});
Object.defineProperty(exports, "useFormApi", {
  enumerable: true,
  get: function () {
    return _index.useFormApi;
  }
});
Object.defineProperty(exports, "useFormState", {
  enumerable: true,
  get: function () {
    return _index.useFormState;
  }
});
Object.defineProperty(exports, "withField", {
  enumerable: true,
  get: function () {
    return _withField.default;
  }
});
Object.defineProperty(exports, "withFormApi", {
  enumerable: true,
  get: function () {
    return _withFormApi.default;
  }
});
Object.defineProperty(exports, "withFormState", {
  enumerable: true,
  get: function () {
    return _withFormState.default;
  }
});
var _baseForm = _interopRequireDefault(require("./baseForm"));
var _arrayField = _interopRequireDefault(require("./arrayField"));
var _index = require("./hooks/index");
var _withField = _interopRequireDefault(require("./hoc/withField"));
var _withFormState = _interopRequireDefault(require("./hoc/withFormState"));
var _withFormApi = _interopRequireDefault(require("./hoc/withFormApi"));
var _interface = require("./interface");
Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interface[key];
    }
  });
});
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
