// @douyinfe/semi-ui/lib/es/configProvider/index.js
var __rest = this && this.__rest || function (s, e) {
  var t = {};
  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};
import React from 'react';
import PropTypes from 'prop-types';
import { BASE_CLASS_PREFIX } from '@douyinfe/semi-foundation/lib/es/base/constants';
import DefaultLocale from '../locale/source/zh_CN';
import Context from './context';
export const ConfigConsumer = Context.Consumer;
export default class ConfigProvider extends React.Component {
  constructor(props) {
    super(props);
  }
  renderChildren() {
    const {
      direction,
      children
    } = this.props;
    if (direction === 'rtl') {
      return /*#__PURE__*/React.createElement("div", {
        className: `${BASE_CLASS_PREFIX}-rtl`
      }, children);
    }
    return children;
  }
  render() {
    const _a = this.props,
      {
        children,
        direction
      } = _a,
      rest = __rest(_a, ["children", "direction"]);
    return /*#__PURE__*/React.createElement(Context.Provider, {
      value: Object.assign({
        direction
      }, rest)
    }, this.renderChildren());
  }
}
ConfigProvider.propTypes = {
  locale: PropTypes.object,
  timeZone: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  getPopupContainer: PropTypes.func,
  direction: PropTypes.oneOf(['ltr', 'rtl'])
};
ConfigProvider.defaultProps = {
  locale: DefaultLocale,
  direction: 'ltr'
};
