// @douyinfe/semi-ui/lib/cjs/markdownRender/components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "a", {
  enumerable: true,
  get: function () {
    return _a.default;
  }
});
Object.defineProperty(exports, "code", {
  enumerable: true,
  get: function () {
    return _code.default;
  }
});
Object.defineProperty(exports, "h1", {
  enumerable: true,
  get: function () {
    return _h.default;
  }
});
Object.defineProperty(exports, "h2", {
  enumerable: true,
  get: function () {
    return _h2.default;
  }
});
Object.defineProperty(exports, "h3", {
  enumerable: true,
  get: function () {
    return _h3.default;
  }
});
Object.defineProperty(exports, "h4", {
  enumerable: true,
  get: function () {
    return _h4.default;
  }
});
Object.defineProperty(exports, "h5", {
  enumerable: true,
  get: function () {
    return _h5.default;
  }
});
Object.defineProperty(exports, "h6", {
  enumerable: true,
  get: function () {
    return _h6.default;
  }
});
Object.defineProperty(exports, "img", {
  enumerable: true,
  get: function () {
    return _img.default;
  }
});
Object.defineProperty(exports, "p", {
  enumerable: true,
  get: function () {
    return _p.default;
  }
});
Object.defineProperty(exports, "table", {
  enumerable: true,
  get: function () {
    return _table.default;
  }
});
var _h = _interopRequireDefault(require("./h1"));
var _h2 = _interopRequireDefault(require("./h2"));
var _h3 = _interopRequireDefault(require("./h3"));
var _h4 = _interopRequireDefault(require("./h4"));
var _h5 = _interopRequireDefault(require("./h5"));
var _h6 = _interopRequireDefault(require("./h6"));
var _a = _interopRequireDefault(require("./a"));
var _img = _interopRequireDefault(require("./img"));
var _table = _interopRequireDefault(require("./table"));
var _p = _interopRequireDefault(require("./p"));
var _code = _interopRequireDefault(require("./code"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
